viewing_party/party.py

```diff
@@ -1,23 +1,165 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    movie = {}
+
+    if title == None or title == "" or genre == None or genre == "" or rating == None or rating == "":
+        return None
+
+    movie["title"] = title
+    movie["genre"] = genre
+    movie["rating"] = rating
+
+    return movie
+
+def add_to_watched(user_data, movie):
+
+    user_data["watched"].append(movie)
+
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+
+    user_data["watchlist"].append(movie)
+
+    return user_data
+
+def watch_movie(user_data, title):
+
+    watchlist = user_data["watchlist"]
+
+    for movie in watchlist:
+        if movie["title"] == title:
+            watchlist.remove(movie)
+            user_data["watched"].append(movie)
+            return user_data
+    
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+
+    if len(user_data["watched"]) == 0:
+        return 0.0
+    
+    sum = 0
+
+    for movie in user_data["watched"]:
+        sum += movie["rating"]
+    
+    return sum / len(user_data["watched"])
+
+
+def get_most_watched_genre(user_data):
+
+    if len(user_data["watched"]) == 0:
+        return None
+
+    genres_count = {}
+
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre in genres_count:
+            genres_count[genre] += 1
+        else:
+            genres_count[genre] = 1
+    
+    max_count = 0
+    most_frequently_watched_genre = ""
+
+    for genre in genres_count.keys():
+        if genres_count[genre] > max_count:
+            max_count = genres_count[genre]
+            most_frequently_watched_genre = genre
+    
+    return most_frequently_watched_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    unique_movies = []
+
+    for movie in user_data["watched"]:
+        is_unique = True
+        for friend in user_data["friends"]:
+            for friend_movie in friend["watched"]:
+                if movie["title"] == friend_movie["title"]:
+                    is_unique = False
+        
+        if is_unique:
+            unique_movies.append(movie)

```

Comment:
I see where y'all were going with this, but we do end up with a triply nested loop, which increases both the code complexity as well as the time and space complexity. 

This is one situation where it's actually better to separate out our separate pieces. We are able to access the user's "watched" list pretty easily (`user_data["watched"]`). What if we first used a doubly nested list to create a list of all the movies watched by the user's friends (We don't care if there are duplicates, because the first match will be returned later)?

Once we've got that list, we can loop through the user's watched list and see if each title exists in the friends movies list we just created. We still end up with two sets of doubly nested lists which is still not the best, but it does prevent us from having a O(n^3).